### ISSUE TITLE
[UX] Add configuration options to search block

### ISSUE THREAD

**Issue by findlabnet** [backdrop/backdrop-issues#2923]
Currently search field haven't placeholder, I think it should be useful for UX.
So, core/search.module line 984 should be
`    '#attributes' => array('title' => t('Enter the terms you wish to search for.'), 'placeholder' => t('Search on site')),`

Search form options
<img width="598" alt="screen shot 2019-01-09 at 4 12 20 pm" src="https://user-images.githubusercontent.com/397895/50937613-63144c00-1429-11e9-93fa-82960c78bf3c.png">

 
---

~PR by @malfroyegon: https://github.com/backdrop/backdrop/pull/2051~
~PR by @findlabnet: https://github.com/backdrop/backdrop/pull/2064~
~PR by @jenlampton at https://github.com/backdrop/backdrop/pull/2071~
New PR by @jenlampton https://github.com/backdrop/backdrop/pull/2468

**Comment by malfroyegon** [comment]
Done :)

**Comment by olafgrabienski** [comment]
Thanks for the PR, @malfroyegon! I've just tested it on the sandbox site, works good.

Regarding the initial request, I like the option to display a placeholder for the search form but I wouldn't be happy to use it on every site, so I'd prefer a respective setting on the search form block.

**Comment by quicksketch** [comment]
I think adding the placeholder text is a good improvement. Making it an option might be a good enhancement for the future, but the simple addition of a hard-coded string seems like an easy starting point.

I like the string itself as well, simple and descriptive.

**Comment by jenlampton** [comment]
I almost always add placeholder text to my search forms, but usually just the string 'Search'. "Search on site" seems like a funny use of language to me, but still an improvement!

**Comment by opi** [comment]
> The placeholder is a short hint intended to aid the user with data entry so it should not be identical to the label element.

From https://www.w3.org/WAI/GL/low-vision-a11y-tf/wiki/Placeholder_Research#Avoid_use_of_placeholder_values

For me having a placeholder value like "Search" or "Search on site" does not describe the expected kind of value. I'd rather go with something short, like "Search terms".

**Comment by findlabnet** [comment]
[PR #2064](https://github.com/backdrop/backdrop/pull/2064) - simple configurable approach for this issue. Field can be just left empty.
Below addition for search settings page

![placeholder](https://user-images.githubusercontent.com/11479969/34639598-3476c590-f2eb-11e7-8496-19400db91362.png)

**Comment by olafgrabienski** [comment]
Had a look at the sandbox site of the new PR. Works as expected.

In my opinion, the configuration option is important because a search block provides already a "Search" heading and a "Search" button by default. Having another "Search whatever" string as a placeholder is quite repetitive, so I'm happy to be able to not use the placeholder, thanks a lot for the PR!

That said, does it make more sense (and is it easy enough to implement) to provide the placeholder setting on the block settings form? Could be useful when placing the search form in different layouts, for instance on the home page without heading but with placeholder, and on other pages viceversa.

**Comment by findlabnet** [comment]
@olafgrabienski, thanks for your opinion. My use case - calling search form from template programmatically and styling it like on screenshot below.

![search](https://user-images.githubusercontent.com/11479969/34651824-6d172aa4-f3de-11e7-8193-f8b856b95197.png)

Maybe someone else can add something? 
Anyway, apply patches for each core update - not a thing I love to do, so we need some solution.

**Comment by olafgrabienski** [comment]
@findlabnet Thanks for adding your use case!

If nobody's up at the moment to implement a placeholder setting on the block settings form, I'd be happy to be able to use your configurable approach instead. (Guess we need some feedback on the PR from a coder perspective.)

**Comment by jenlampton** [comment]
> Anyway, apply patches for each core update - not a thing I love to do, so we need some solution.

@findlabnet it's the best way to make sure those issues get fixed in core (and contrib, too!) and it's not painful if you document everything.  See http://www.jenlampton.com/blog/sustainable-development-workflow-patching for how I manage it :)

**Comment by findlabnet** [comment]
@jenlampton, thanks for useful link. I thought that we are different from a very bureaucratic D-world and can do useful things much faster.

**Comment by quicksketch** [comment]
I feel like a single option for Placeholder text seems out of place when no other aspects of the form are configurable. Take @findlabnet's excellent example from the screenshot above. I think that's a typical configuration of the search form: No label (or a hidden label), no button (but usually a search icon), and a placeholder. In my experience that's actually the most common display of a search form. Anecdotally, here on GitHub, that's exactly what they do.

My feeling is that if we make placeholder configurable, so should the button text and the label (primarily so that they could be disabled).

Like @olafgrabienski, I think the most expected place for these options would be on the search block form, not the global configuration.

> My use case - calling search form from template programmatically and styling it like on screenshot below (no heading, no button text).

If the search form is called from via the API, there is already the ability to modify any aspect, including the placeholder, right? What's the need for an option at all in this use-case?

**Comment by jenlampton** [comment]
> What's the need for an option at all in this use-case?

For people who are not experienced enough yet to use the API ;)

>  rename this one from "Add placeholder to search field" to something like "Add configuration options to search block"?

Yes please :)

**Comment by quicksketch** [comment]
Nice, I really like the direction here. I also think newly placed search blocks should default to the block title being disabled. Search forms are usually clear about their purpose by their appearance, and there's no need to include a block title in most cases.

Despite this being a neat feature, it's too late for 1.9.0. We'll need more feedback and tests added to validate these new block settings.

**Comment by jenlampton** [comment]
Changing from NR to NW because we need a checkbox for "Show submit button" instead of asking the user to leave that field blank to disable.

**Comment by jenlampton** [comment]
Okay, pushed a new PR that Adds more options and uses `#states` to show/hide dependant fields.

No options selected:
![screen shot 2018-04-30 at 5 40 37 pm](https://user-images.githubusercontent.com/397895/39456787-e68406d8-4c9d-11e8-8600-7d52a5f78f92.png)

All options selected: 
![screen shot 2018-04-30 at 5 41 03 pm](https://user-images.githubusercontent.com/397895/39456792-ed5faa0c-4c9d-11e8-9479-22f9b89b2eae.png)

**Comment by stpaultim** [comment]
Trying to be helpful, but not 100% certain I understand the process or expectations. 
I used the sandbox provided here to test the functionality of the search block. Things generally seem to work as I understand they are supposed to work. 
https://github.com/backdrop/backdrop/pull/2071

**Comment by stpaultim** [comment]
I am also a little confused about the use case for NOT displaying the search button?

So there are at least two options that I don't quite understand why they are there. But, they work as expected. An argument can be made that options without a solid use case are a distraction or possible UX problem. But, I'm assuming that others have use cases that I simply haven't thought of yet.

There seems to be a switch from search form to search box. I am not seeing any discussion about why it's Search Box and not Search Block. I don't have a problem with that as long as it consistent with other blocks.

**Comment by jenlampton** [comment]
>  I confused the label with the title

The label is the form element label, and is useful for screen readers so that people know what to enter into the input field. 

The title is the block title. That's an option that exists for all blocks, and is hard to remove from this form, but is irrelevant to the specific search form options -- other than it makes the whole thing really confusing :)

In most cases there won't be a title.

**Comment by jenlampton** [comment]
> I am also a little confused about the use case for NOT displaying the search button?

It's most common that people put a magnifying glass icon inside the search box (with CSS) instead of using a button. If you do a [google image search](https://www.google.com/search?q=search+box&rlz=1C5CHFA_enUS724US724&source=lnms&tbm=isch&sa=X&ved=0ahUKEwiB1Ovsu-PaAhVL34MKHWICDj8Q_AUICigB&biw=1470&bih=944) for "search box" you'll see what I mean. 

> There seems to be a switch from search form to search box.

I did this as an attempt to make the language match what humans are thinking about when adding a search box to a page. 

> I am not seeing any discussion about why it's Search Box and not Search Block.

There wasn't a discussion, but we can have one now :)

* Only people familiar with Backdrop / Drupal know what a block is. Everyone knows what a search box is.
* These new settings are specifically for the box where you enter text. Should that box have a label, placeholder text, or be followed by a button? 
* I thought using "Search box" might clear up some of the confusion from title vs label  - since the title is for the block. But I was Obviously wrong about that ;)

**Comment by stpaultim** [comment]
The case for not displaying a search button became clear as I played with it and reflected on sites I have built in the past. But, when I turned off the search button and then visited the page, the first thought that struck me was: "That's not very helpful. Now, I can't activate the search." This was probably because only seconds before I was looking at the search with a button and had that image fresh in my head. Also, hitting return/enter to get an action always ma"kes me nervous, because sometimes hitting return/enter can trigger the wrong action. But, that's why we are giving site builders the option of using a button or not. 

A site builder might be confused for a minute, like I was. But they will figure it out and will appreciate having the control. I would not call this a problem. 

I see the case for calling it a search box. I may be too ingrained in Drupal/Backdrop culture, because "box" sounds a little weird to me. I also was a little surprised, because I see other blocks called "Hero Block" and "Header Block." Switching to box felt inconsistent. But, most of the pre-configured blocks do not have "block" in their name and in fact upon reflection, I can see that it's redundant. It's the block with the search box. Hero block should probably be the block with the hero image (or something like that). I like it.

**Comment by jenlampton** [comment]
Naming things is hard ;)

**Comment by olafgrabienski** [comment]
> The label is the form element label specifically for the search box (...) and is most useful for screen readers so that people know what to enter into the search box

I've two ideas regarding this topic:

- Add a respective help text to the configuration form.
- Provide the label element with a screenreader-only class so that it's hidden for people who don't use a screenreader. (Do we already have such a class?)

**Comment by olafgrabienski** [comment]
> Here's a link to the latest PR for testing ...

Have tested the configuration options via the sandbox site and I've the impression that everything is working fine!

**Comment by olafgrabienski** [comment]
I'd say RTBC but I'm not able to make a code review, so feel free to decide otherwise.

**Comment by olafgrabienski** [comment]
>> Provide the label element with a screenreader-only class so that it's hidden for people who don't use a screenreader.

> We are getting into edge-cases here for people who want a custom label but also want it visible only to screen-readers. Let's leave this to contrib.

@jenlampton - your explanation ("most useful for screen readers") didn't sound like edge-case.

**Comment by jenlampton** [comment]
@olafgrabienski the edge case was where people wanted it hidden. Screen readers can read visible text without any issue!

**Comment by olafgrabienski** [comment]
> the edge case was where people wanted it [the label] hidden too

Hm, according to @quicksketch's comment above -- https://github.com/backdrop/backdrop-issues/issues/2923#issuecomment-357547551 --, the "typical configuration" and "most common display" is: no label or a hidden one (better for screenreaders).

I agree with the linked comment, and that's why I still don't see my proposal ("Provide the label element with a screenreader-only class so that it's hidden for people who don't use a screenreader") as an edge case.

**Comment by quicksketch** [comment]
I did some background research to see if hidden labels are still "a thing", and thus far it looks like hidden labels are still very common and in many cases required. Some places where there is only one field (e.g. Google) do not include a hidden label, but sites where there is a lot more going on (e.g. Amazon) do include a hidden label on their search fields.

Tests are still failing on this PR, and I think @olafgrabienski is right that we should always print the label, but use `#title_display = 'invisible'` rather than not printing it all.

**Comment by jenlampton** [comment]
So, are we then trying to support the edge-case of a customized label that's also hidden? Or can we use the label "Search" when the label isn't specified?

**Comment by olafgrabienski** [comment]
> So, are we then trying to support the edge-case of a customized label that's also hidden? Or can we use the label "Search" when the label isn't specified?

As a hidden label seems to be quite common, I still don't understand why customizing it turns the case into an edge-case. Isn't it more straightforward to treat the label always the same, regardless of its visibility?

**Comment by jenlampton** [comment]
> Isn't it more straightforward to treat the label always the same, regardless of its visibility?

Yes, if you understand why. However, I'm betting most people will be confused by why they need to add a label if they can't see it themselves. Maybe some help text can clear that up...

I'm taking a guess at the next steps for this PR:
* get tests passing
* make the label field required
* add help text explaining why

**Comment by jenlampton** [comment]
Okay, I've pushed an updated PR that I believe addresses these issues.

**Comment by opi** [comment]
Screenshot for the block setting modal, according to last PR by jen:
![capture d ecran de 2019-01-04 09-07-54](https://user-images.githubusercontent.com/159548/50678729-7033d400-1000-11e9-8254-b2d6abd7c082.png)

I'd like these settings to be wrapper into an open fieldset, called "Search form settings". @jenlampton agree ? Can you add it to your PR, otherwise I'll do.

**Comment by jenlampton** [comment]
yes, that's much better :)

**Comment by jenlampton** [comment]
This is what I was thinking initially, but we also need to account for the use-case of having a customized label that's hidden.

**Comment by docwilmot** [comment]
Ok, so that means no states, but the description can still go.

**Comment by jenlampton** [comment]
fair enough, removed :) https://github.com/backdrop/backdrop/pull/2468

**Comment by jenlampton** [comment]
> Little feedback left.

Changes made, pr updated.

> I'm also thinking we should hide the block title by default.

I completely agree, but I don't think we have a clean way to change a (layout) block setting from within the (search) block configure form. Maybe with a form alter? Trying that...

**Comment by jenlampton** [comment]
> Maybe with a form alter? Trying that...

Well, that get's it done. But it feels like a bit of a hack. Maybe Layout module should add a `#block`, `#delta`, and a '#new` for a cleaner form-alters?

**Comment by jenlampton** [comment]
Okay...  I knew they had to be there. Helps if I look in the right place :) This feels much better.

```php
/**
 * Implements hook_form_FORM_ID_alter().
 */
function search_form_layout_block_configure_form_alter(&$form, &$form_state) {
  // Only alter the search block's configure form.
  if ($form_state['block']->module == 'search' && $form_state['block']->delta == 'form') {
    // Only hide the block title when the search block is first added.
    if ($form_state['block']->is_new == TRUE) {
      $form['title_display']['title_display']['#default_value'] = 'none';
    }
  }
}
```

**Comment by findlabnet** [comment]
More than year ago we can had two or three simple working, but not so perfect solutions. 
In the meantime, we have nothing yet but the path to perfection.

Maybe we can release things by more iterative way, something better only after something working?

This is more common question, not only for this particular issue, we can all see more examples for very long approaches to achieve very simple things.

Any opinions about acceptability such workflow duration?

**Comment by docwilmot** [comment]
@findlabnet I agree its annoying when things are slow. I don't think it needs to be perfect but it still needs to be acceptable, and I think every discussion in this issue has been very reasonable thoughts and concerns. We are close now though.

**Comment by jenlampton** [comment]
@findlabnet Thanks for the reminder. 

I think on lots of issues we are pretty good about getting it done first, and then making it better later, but we do constantly have to fight the urge for perfection, and reminders help! You're right, this solution has been lagging for quite a while, but @docwilmot is also right that we're really close. 

If you want to help get this done, all it needs is a review! Please [log in to the sandbox](https://github.com/backdrop/backdrop/pull/2468#issuecomment-453373974), and add a search block to the site. If you like where this solution is right now, you can change the label on this issue from "pr - needs review" to "pr - reviewed and tested by the community" and that will increase the likelihood of getting this into 1.12!

**Comment by findlabnet** [comment]
@jenlampton I glad to help, when I can. See my comment to PR.

**Comment by jenlampton** [comment]
Thanks for that @findlabnet 👍  I have pushed a fix to the PR and it is ready for testing again. :)

**Comment by findlabnet** [comment]
See my comment to PR. If it's OK, status can be changed to "pr - reviewed and tested by the community".

**Comment by jenlampton** [comment]
> See my comment to PR

Another good one. I have pushed one more (last?) fix to the PR and it is ready for (final?) testing. :)

**Comment by findlabnet** [comment]
Seems good, thanks to all involved.

**Comment by quicksketch** [comment]
I've merged https://github.com/backdrop/backdrop/pull/2468 into 1.x for 1.12.

Thanks @findlabnet, @malfroyegon, @jenlampton, @olafgrabienski, @stpaultim, @docwilmot, and @opi!
